### ISSUE TITLE
Unable to add translation for dataobjects.

### ISSUE THREAD

**Issue by letsbuildafire** [silverstripe/silverstripe-translatable#220]
Problem is like [Issue 196](https://github.com/silverstripe/silverstripe-translatable/issues/196).

Available translation list is populated correctly and button is displayed but clicking on it does nothing (no javascript errors, network requests).

Currently using SS 3.2.1 and the Translatable module and that's it, composer looks like this:

``` json
{
    "name": "silverstripe/installer",
    "description": "The SilverStripe Framework Installer",
    "require": {
        "php": ">=5.3.3",
        "silverstripe/cms": "3.2.1",
        "silverstripe/framework": "3.2.1",
        "silverstripe/reports": "3.2.1",
        "silverstripe/siteconfig": "3.2.1",
        "silverstripe/translatable": "^2.1"
    },
    "require-dev": {
        "phpunit/PHPUnit": "~3.7"
    },
    "config": {
        "process-timeout": 600
    },
    "prefer-stable": true,
    "minimum-stability": "dev"
}
```

and my dataobject begins like:

``` php
<?php

class Format extends DataObject implements PermissionProvider {
    private static $extensions = array(
        "Translatable"
    );

    private static $db = array(
        'Title'=>'Varchar(128)',
    );
        ...
```

**Comment by tractorcow** [comment]
Hm, interesting. So it's the inline action that's not working correctly? I wonder if that feature ever worked for non-page dataobjects.

In fact, I'm not sure what mechanisms are in place to assist with non-page translation.

How are you managing this object? Is it inside a model admin, or inside a gridfield and attached to a parent object? If the latter, what kind of relationship?

**Comment by letsbuildafire** [comment]
Yeah, it appears to be the inline action not working. Though I did also attempt changing it to just a regular FormAction (as mentioned in Issue 196) to no avail.

The object is in a model admin, just a single dataobject per admin class at the moment.

**Comment by tractorcow** [comment]
It looks like TranslatableCMSMainExtension is applied to all leftandmain...

So I guess we should start by figuring out why the button doesn't issue a postback, and work from there.

**Comment by letsbuildafire** [comment]
Just tested a couple of other things. It looks like descendants of SiteTree cannot be translated when managed by a model admin, though work correctly when managed through the SiteTree.

**Comment by christopherdarling** [comment]
We've run into something very similar to this on a project, Translateable works fine via SiteTree pages, we have multiple DataObject's managed by GridField on those pages which we've applied the Translatable extension to, the tab appears in the CMS when editing the DataObject's but nothing happens on click of the 'Create' button.

This looks to be simply down to the selector used ('.LeftAndMain :input[name=action_createtranslation]') in CMSMain.Translateable.js changing it temporarily to :input[name=action_createtranslation] applies the onclick handler but we get a 404 error in the Network panel on Dev Tools. The 404 looks to be because it can't find the createtranslation action.

Adding TranslatableCMSMainExtension extension to GridFieldDetailForm_ItemRequest via config seems to work and stop the 404 error but that now results in a 500 error. This is due to https://github.com/silverstripe/silverstripe-translatable/blob/2.1/code/controller/TranslatableCMSMainExtension.php#L129-L131 which I've managed to get around by changing to
```php
	if ( $this->owner instanceof GridFieldDetailForm_ItemRequest ) {
		$controller = $this->owner->getController();
	} else {
		$controller = $this->owner;
	}
	$controller->getResponse()->addHeader('X-Pjax', 'Content');

	return $controller->redirect($url);
```

That leaves us with the incorrect URL in the redirect (admin/pages/edit/EditForm/field/DataObjectName/item/1/show/3?Locale=en_GB in my case) which should be admin/pages/edit/EditForm/field/StructuralTypes/item/3?Locale=en_GB which I've managed to resolve but the ?Locale= param remains as the original language not the newly translated one

I've pushed up the changes to my fork here incase it's of any help; https://github.com/christopherdarling/silverstripe-translatable/tree/gridfieldfix

**Comment by tractorcow** [comment]
Would you like to put it into a pull request? Even if it doesn't solve all the issues, it seems like an improvement.

**Comment by christopherdarling** [comment]
Sure; done here https://github.com/silverstripe/silverstripe-translatable/pull/235

**Comment by robbieaverill** [comment]
My (quickly assembled) assumption is that this is fixed with #238, please feel free to reopen if I'm wrong :)
